README.md
@sailrs/pwa
===========

`@sailrs/pwa` leverages google's workbox and exposes its API in a more
convenient way. Pretty much like with workbox itself, you can either use
`recipes` or configure your service worker completely as you desire.

## Using Recipes

The easiest way to configure your service worker the following snippet:

```typescript
import { setupServiceWorker } from '@sailrs/pwa'

setupServiceWorker({
  recipes: {
    pageCache: true,
    imageCache: true,
    staticResourcesCache: true,
    offlineFallback: {
      pageFallback: 'my-offline-page',
      imageFallback: '/assets/fallback.jpg',
    },
    warmCache: {
      urls: [
        '/join',
        '/login',
      ],
    }
  },
})
```

Except the google fonts cache, this snippet sets up all recipes from the
workbox documentation. We take it one step further,  because `@sailrs/pwa`
enabled you to compose your caching needs which are enabled here too, where
appropriate. For instance, you cannot change the strategy for `pageCache` but
all the options, that the NetworkFirst Strategy exposes:

```typescript
recipes: {
  pageCache: {
    matchOptions: {
      ignoreSearch: true,
      ignoreVary: true,
    },
    fetchOptions: {
      credentials: 'include',
    },
    networkTimeoutSeconds: 3,
  },
}
```

On the other hand, for the `imageCache` it may make sense to alter the
`matchCallback` or the plugins.

```typescript
imageCache: {
  match: ({ request }) => request.destination === 'image',
  expiration: {
    maxEntries: 100,
    maxAgeSeconds: 60 * 60 * 24 * 7, // 7 days
  },
  cacheableResponses: {
    statuses: [0, 200],
  }
},
```

On the `warmCache` you can also change the strategy and plugins.

```typescript
warmCache: {
  strategy: 'staleWhileRevalidate',
  urls: [
    '/join',
    '/login',
  ],
  expiration: {
    maxEntries: 100,
    maxAgeSeconds: 60 * 60 * 24 * 7, // 7 days
  },
  cacheableResponses: {
    statuses: [0, 200],
  },
},
```

## Custom Caching Strategies

Use the `caches` section of the configuration to compose the caches as you like.
The key becomes the name of the cache.

```typescript
const config: Config = {
  caches: {
    images: {
      match: /\.(?:png|gif|jpg|jpeg|svg)$/,
      strategy: 'staleWhileRevalidate',
      expiration: {
        maxEntries: 100,
        maxAgeSeconds: 30 * 24 * 60 * 60, // 30 Days
      },
      ...
    }
  },
}
```

## Handling service worker updates

When a new service worker is released

```typescript
export function promptForUpdate(): Promise<boolean> {
  return new Promise((res) => res(confirm("Update available. Reload now?")));
}
```


CHANGELOG.md


### 0.3.3 (2023-07-20)


### Bug Fixes

* await showSkipWaitingPrompt ([0f5d5b5](https://github.com/sailrs-io/pwa/commit/0f5d5b58b5b3d5c4697c853f17cfa932953ce0cd))

# Changelog

### 0.3.2 (2023-04-20)

### Bug Fixes

- export submodules ([174026c](https://github.com/sailrs-io/pwa/commit/174026ce27ee2c244144d1809e003c7e65356efc))

### [0.3.1](https://github.com/sailrs-io/pwa/compare/v0.3.0...v0.3.1) (2023-04-20)

## [0.3.0](https://github.com/sailrs-io/pwa/compare/v0.2.2...v0.3.0) (2023-04-17)

### Features

- implemented recipes and custom caches ([0ade343](https://github.com/sailrs-io/pwa/commit/0ade3439afb20889ce3b44fd242ab5641bdcf912))
- **notifications:** ask for permission and handle users' choice ([7780d11](https://github.com/sailrs-io/pwa/commit/7780d11ecf14a6195619ed9cce78b5b179a27a0f))
- **notifications:** provide default event handlers ([8b2309b](https://github.com/sailrs-io/pwa/commit/8b2309b3741c4cae98e2a982a2eba3afbb4c1d9b))
- **push:** register a push event handler ([390e938](https://github.com/sailrs-io/pwa/commit/390e9381f17b77ac531050408e9274e035b1bb3b))
- **update:** ask user to reload page and send skipWaiting to worker ([2dae88c](https://github.com/sailrs-io/pwa/commit/2dae88c1903ec310e435b3bb136d260d251283a4))

### Bug Fixes

- bundling with esbuild ([7d68240](https://github.com/sailrs-io/pwa/commit/7d682409acc398a6713e779dcc5830eca466faca))
- imports ([d3192a4](https://github.com/sailrs-io/pwa/commit/d3192a407bbba5483454eca3d9c7156e54463af0))
- workflows ([93e72aa](https://github.com/sailrs-io/pwa/commit/93e72aa50ce063ed48f86242a5e55960e3b301a0))

src/lib/types/routes.ts
import { OfflineFallbackOptions } from "workbox-recipes"
import { WarmCacheArgs } from "../cache/recipes/warmCache.js"
import { StrategyConfig } from "./strategies.js"
import { NetworkFirstOptions } from "workbox-strategies"
import { StaticResourcesCacheArgs } from "../cache/recipes/staticResourcesCache.js"
import { ImageCacheArgs } from "../cache/recipes/imageCache.js"

export type RouteConfig = {
  [key: string]: StrategyConfig
}

export type RecipeConfig = {
  /**
   * imageCache recipe
   * ```js`
   * config: {
   *   match: ({ request }) => request.destination === 'image',
   *   strategy: 'staleWhileRevalidate',
   *   cacheableResponses: {
   *     statuses: [0, 200]
   *   },
   *   expiration: {
   *     maxAgeSeconds: 30 * 24 * 60 * 60, // 60 days
   *     maxEntries: 60
   *   },
   * }
   * ```
   *
   * @see https://developer.chrome.com/docs/workbox/modules/workbox-recipes/#image-cache
   */
  imageCache?: Boolean | ImageCacheArgs

  /**
   * offlineFallback recipe
   *
   * @param pageFallback - The URL of the page to show when the user is offline.
   * @param imageFallback - The URL of the image to show when the user is offline.
   * @param fontFallback - The URL of the font to show when the user is offline.
   *
   * @see https://developer.chrome.com/docs/workbox/modules/workbox-recipes/#offline-fallback
   */
  offlineFallback?: OfflineFallbackOptions

  /**
   * pageCache recipes 
   * ```js
   * config: {
   *   match: ({request}) => request.mode === 'navigate',
   *   strategy: 'networkFirst',
   *   networkTimeoutSeconds: 3,
   *   cacheableResponses: {
   *     statuses: [0, 200]
   *   },
   * }
   * ```
   *
   * @see https://developer.chrome.com/docs/workbox/modules/workbox-recipes/#page-cache
   */
  pageCache?: Boolean | NetworkFirstOptions

  /** 
   * staticResourcesCache recipe
   *
   * ```js
   * config: {
   *   match: ({request}) => ['style', 'script', 'worker'].includes(request.destination),
   *   strategy: 'staleWhileRevalidate',
   *   cacheableResponses: {
   *     statuses: [0, 200]
   *   },
   * }
   * ```
   *
   * @see https://developer.chrome.com/docs/workbox/modules/workbox-recipes/#static-resources-cache
   */
  staticResourcesCache?: Boolean | StaticResourcesCacheArgs

  /** 
   * warmCache recipe
   * ```js
   * config: {
   *   strategy: 'cacheFirst',
   *   urls: [],
   * }
   * ```
   *
   * @see https://developer.chrome.com/docs/workbox/modules/workbox-recipes/#static-resources-cache
   */
  warmCache?: WarmCacheArgs
}

export type Config = {
  /** configure your custom caching needs  */
  caches?: RouteConfig
  /** use preconfigured workbox recipes */
  recipes?: RecipeConfig
  /** handle push notifications  */
  pushNotifications?: {
    /** handler function that receives the PushEvent */
    onPush: (event: PushEvent) => Promise<void>
  }
}
